### ISSUE TITLE
glfwSetWindowAttrib() for GLFW_MOUSE_PASSTHROUGH causes glfwWaitEvents() to return, as if there's a new event (WINDOWS 11)

### ISSUE THREAD

**Issue by MirrasHue** [glfw/glfw#2600]
Hey, I'm experimenting with GLFW and ImGui. What I have now is glfwWaitEvents() on the main thread, and update and rendering in a separate thread, it works great for not blocking the execution while resizing, moving the window or holding the minimize, maximize or close buttons.

I noticed that while rendering ImGui, glfwWaitEvents() never waits, but exits immediately, even when I'm not providing any inputs.

Upon some investigation, I found the problem: ImGui is calling glfwSetWindowAttrib() every frame, and this function calls SetWindowLongW(), which seems to be emitting an unknown event (couldn't catch it in any callback), causing glfwWaitEvents() to not wait for user input.

Taking a look at the implementation of ImGui_ImplGlfw_UpdateMouseData (which calls glfwSetWindowAttrib), there's no check to see if the mouse passthrough attribute is already set to the desired state (enabled / disabled).

I know this check could also be implemented in _glfwSetWindowMousePassthroughWin32() itself, but this event being fired by SetWindowLongW() got me curious.

**Comment by dougbinks** [comment]
I am using `glfwWaitEvents` with Dear ImGui and have not observed this on Windows 11 - although `glfwSetWindowAttrib` is being called every frame it isn't causing any events to be generated, and `glfwWaitEvents` doesn't return.

I am wondering if this is caused by some third party DLL monitoring the window changes - it might be worth inspecting your running exe for attached DLLs, which you can do by adding a breakpoint in Visual Studio and  debug modules window (sorted by path to look for any hooked DLLs).

**Comment by MirrasHue** [comment]
Hi @dougbinks, thanks for the reply.

Did you move your rendering to a separate thread?

I'll try to provide a simple minimal example:

```cpp
#include <stdio.h>
#include <thread>

#include <glad/glad.h>
#include <glfw/glfw3.h>

#include <imgui_impl_glfw.h>
#include <imgui_impl_opengl3.h>
#include <imgui.h>

static GLFWwindow* windowHandle = nullptr;
static bool running = true;

void update()
{
    glfwMakeContextCurrent(windowHandle);

    while(running)
    {
        glClearColor(0.1f, 0.1f, 0.1f, 1.f);
        glClear(GL_COLOR_BUFFER_BIT);
        
        ImGui_ImplOpenGL3_NewFrame();
        ImGui_ImplGlfw_NewFrame();
        ImGui::NewFrame();

        ImGui::ShowDemoWindow();

        ImGui::Render();

        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());

        glfwSwapBuffers(windowHandle);
    }
}

void run()
{
    glfwMakeContextCurrent(nullptr);

    std::jthread updateThread{update};

    while(running)
    {
        glfwWaitEvents();
        printf("This shouldn't be printed if we aren't doing anything\n");
    }
}

int main()
{
    glfwInit();

    windowHandle = glfwCreateWindow(1280, 720, "ImGui", nullptr, nullptr);
    glfwMakeContextCurrent(windowHandle);
    glfwSwapInterval(1);

    gladLoadGL(glfwGetProcAddress);

    glfwSetWindowCloseCallback(windowHandle, [](GLFWwindow* window){
        running = false;
    });

    // Setup Dear ImGui context
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO();

    io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard; // Enable Keyboard Controls
    io.ConfigFlags |= ImGuiConfigFlags_DockingEnable;     // Enable Docking

    // Setup Dear ImGui style
    ImGui::StyleColorsDark();

    // Setup Platform/Renderer backends
    ImGui_ImplGlfw_InitForOpenGL(windowHandle, true);
    ImGui_ImplOpenGL3_Init("#version 330");

    run();    
}
```

In my case, the message is being printed all the time. This is what I did to fix it:

On imgui_impl_glfw.cpp

```cpp
#if GLFW_HAS_MOUSE_PASSTHROUGH
        const bool window_no_input = (viewport->Flags & ImGuiViewportFlags_NoInputs) != 0;
        //glfwSetWindowAttrib(window, GLFW_MOUSE_PASSTHROUGH, window_no_input);

        // Bug fix (for multithreading), don't call glfwSetWindowAttrib with GLFW_MOUSE_PASSTHROUGH every frame, because it ends up firing an
        // unknown event, which makes glfwWaitEvents (on the main thread) to not wait, and causes problems when synchronizing the window resize
        bool has_attribute = glfwGetWindowAttrib(window, GLFW_MOUSE_PASSTHROUGH);

        if(window_no_input && !has_attribute)
            glfwSetWindowAttrib(window, GLFW_MOUSE_PASSTHROUGH, true);
        else if(!window_no_input && has_attribute)
            glfwSetWindowAttrib(window, GLFW_MOUSE_PASSTHROUGH, false);
#endif
```

About the more advanced debugging tools you mentioned, I'm using vscode with gbd, which probably don't have those capabilities

**Comment by dougbinks** [comment]
> About the more advanced debugging tools you mentioned, I'm using vscode with gbd, which probably don't have those capabilities

You can [download and use Process Explorer from Microsoft](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer) to view the DLLs your exe has loaded.

> Did you move your rendering to a separate thread?

No, I use `glfwWaitEvents` to stop rendering when there is no user input, and in my case this works well. If the event is sent after `glfwSetWindowAttrib` then `glfwWaitEvents` would not wait. It's possible there's something slightly more complex going on though, such as the event being generated when the window is updated (rendered to) after the attribute change, which could explain the issue.

It's likely worth adding a check before set to GLFW though.

**Comment by dougbinks** [comment]
Note that I have earlier proposed that `glfwWaitEvents` [might ignore non-GLFW events](https://github.com/glfw/glfw/issues/685#issuecomment-416167614), however this could break some people's code so would likely require an extra attribute.

**Comment by MirrasHue** [comment]
I'll take a look later.

Please let me know if you could run the example I provided. If your compiler doesn't support C++ 20, let me know as well, so that I can edit it to use std:: thread, requiring only C++ 11.

Interesting to see you and other people also experienced this problem.

**Comment by dougbinks** [comment]
I likely won't have time to run the example.

**Comment by coderedart** [comment]
I think this issue may be out of scope because glfw only supports calling `glfwSetWindowAttrib` on main-thread, while your example calls it on rendering thread (not main).

**Comment by MirrasHue** [comment]
Yeah, I suspected that. I ended up modifying imgui so that this function doesn't get called every frame, now it's working fine. Still very intriguing what causes the waiting to be interrupted.

**Comment by dougbinks** [comment]
I have a potential prototype implementation which only exits a wait if a GLFW relevant event is received. This is on my branch: https://github.com/dougbinks/glfw/tree/wait-events

This only has Win32 and Wayland implementations for now, but the other backends should work as before. I've not fully tested all paths for events so some may not cause a wait to end, so I would not advise using this in production for now. However I would appreciate any testing.

This may not fix your issue if `glfwSetWindowAttrib` is causing an actual GLFW event to occur, but as far as I can tell it shouldn't.

**Comment by MirrasHue** [comment]
Hey doug, thanks for taking a look at this.

I suppose this is an actual problem only if calling it from another thread, and in my situation it's not feasible to isolate ImGui (which is the one calling the function) to run on the main thread.

At the moment I can't try it, but I took a look at the code, and it seems to filter and accept only the events recognized by GLFW.

> This may not fix your issue if glfwSetWindowAttrib is causing an actual GLFW event to occur, but as far as I can tell it shouldn't.

Yeah, I think this is not the case, because as I mentioned before, I subscribed to every callback, and none of them fired.

**Comment by MirrasHue** [comment]
coderedart said:

> I think this issue may be out of scope because glfw only supports calling glfwSetWindowAttrib on main-thread, while your example calls it on rendering thread (not main).

Are these thread safety suggestions theoretical, or based only in some OS (which end up dragging all the others down in the documentation)?

It seems that you can get away with a lot of things on Windows, as it doesn't care if you render UI from another thread. Also, elmindreda recommended that we should use multithreading to solve the freezes when manipulating the window (on Windows), so if we worry too much about which functions can be called or not, it's going to be harder to achieve the desired results

**Comment by MirrasHue** [comment]
The example I provided before could be a good way to see if you also had the problem, and if so, test if these changes fix it

**Comment by dougbinks** [comment]
> Are these thread safety suggestions theoretical, or based only in some OS (which end up dragging all the others down in the documentation)?

> It seems that you can get away with a lot of things on Windows, as it doesn't care if you render UI from another thread. Also, elmindreda recommended that we should use multithreading to solve the freezes when manipulating the window (on Windows), so if we worry too much about which functions can be called or not, it's going to be harder to achieve the desired results

The problem here is that GLFW does not have any thread guarantees for `glfwSetWindowAttrib` and the rest of the code. So calling `glfwSetWindowAttrib` from a non-main thread could result in unexpected behaviour.

If you want to multi-thread the rendering the correct way to handle GLFW functions which can only run on the main thread is to pass the information to/from the main thread somehow and then call the functions there.

For example using my [enkiTS](https://github.com/dougbinks/enkits) tasking library I would create a pinned task on the main thread (thread 0) which calls these functions, and make sure the main thread loop calls `RunPinnedTasks()`. I hope to make an example GFLW multithread app with enkiTS at some point.

**Comment by MirrasHue** [comment]
> If you want to multi-thread the rendering the correct way to handle GLFW functions which can only run on the main thread is to pass the information to/from the main thread somehow and then call the functions there.

> For example using my [enkiTS](https://github.com/dougbinks/enkits) tasking library I would create a pinned task on the main thread (thread 0) which calls these functions, and make sure the main thread loop calls RunPinnedTasks(). I hope to make an example GFLW multithread app with enkiTS at some point.

I was familiar with this approach, I saw it on SDL, it was introduced in version 3.

In my case, I'm using wait events on the main thread, so I would need to poll if I wanted to use this (big no no, unnecessary extra CPU usage if there are no events), or use postEmptyEvent() to make it unblock and somehow figure out how to make it reliably execute the queued functions (oh I just realized, the main thread would never sleep anyway, because in every frame ImGui calls functions that are not "thread safe"). 

Another con is that I would have to modify ImGui, so that every non thread safe GLFW function call would have to be synchronized, given that I need the result to proceed right then and there (more overhead)

As my multithreaded approach is only meant to be used on Windows (Linux doesn't have this window problem, macOS has it when I tested version 13, but I'm not focusing on it now), I don't need to worry about these functions, they seem to return the correct results

**Comment by dougbinks** [comment]
> In my case, I'm using wait events on the main thread, so I would need to poll if I wanted to use this (big no no, unnecessary extra CPU usage if there are no events), or use postEmptyEvent() to make it unblock and somehow figure out how to make it reliably execute the queued functions (oh I just realized, the main thread would never sleep anyway, because in every frame ImGui calls functions that are not "thread safe").

Your main thread would sleep most of the time, wake up due to a `postEmptyEvent()` and then process the queued functions.

> Another con is that I would have to modify ImGui, so that every non thread safe GLFW function call would have to be synchronized, given that I need the result to proceed right then and there (more overhead)

Not all the functions need to be synchronized that precisely, however a redesign of ImGui's GLFW backend to be thread safe would be a good idea which I might look at some day.

> As my multithreaded approach is only meant to be used on Windows (Linux doesn't have this window problem, macOS has it when I tested version 13, but I'm not focusing on it now), I don't need to worry about these functions, they seem to return the correct results

One of the issues with multithreading is that thread memory access issues often work most of the time.
